middleware/2fa.js

```diff
@@ -0,0 +1,287 @@
+const nearAPI = require('near-api-js');
+const crypto = require('crypto');
+const nacl = require('tweetnacl');
+const { creatorKeyJson } = require('./near');
+const { sendSms } = require('../utils/sms');
+const { sendMail } = require('../utils/email');
+const models = require('../models');
+const Sequelize = require('sequelize');
+const Op = Sequelize.Op;
+const password = require('secure-random-password');
+const SECURITY_CODE_DIGITS = 6;
+const twoFactorMethods = ['2fa-email', '2fa-phone'];
+const viewMethods = [];
+const changeMethods = ['confirm'];
+
+const getContract = async (contractName, secretKey) => {
+    const keyStore = {
+        async getKey() {
+            return nearAPI.KeyPair.fromString(secretKey);
+        },
+    };
+    const near = await nearAPI.connect({
+        deps: { keyStore },
+        masterAccount: creatorKeyJson && creatorKeyJson.account_id,
+        nodeUrl: process.env.NODE_URL
+    });
+    const contractAccount = new nearAPI.Account(near.connection, contractName);
+    const contract = new nearAPI.Contract(contractAccount, contractName, {
+        viewMethods,
+        changeMethods,
+    });
+    return contract;
+};
+const confirmRequest = async (accountId, request_id) => {
+    const key = await getDetermKey(accountId);
+    const contract = await getContract(accountId, key.secretKey);
+    // always parseInt on requestId. requestId is string in db because it could come from url params etc...
+    const res = await contract.confirm({ request_id: parseInt(request_id) }).catch((e) => {
```

Comment:
why use `.catch` here instead of using `try` / `catch` block?

Using `.then` / `.catch` in `async` function is confusing. I think there is already mistake here:

`return { success: false, error: e };` will effectively return result of the promise to `const res`. So `return { success: !!res, res };` will return `{ success: true, { success: false, error: e}}`. Which suggests that error condition wasn't tested here.

I think it's generally best to stick to `try` / `catch` in `async` functions and let most exceptions fall through until route middleware. Like there is no reason to return error as successful HTTP response with `success: false` when appropriate HTTP error can be returned instead (e.g. 401 in this case).

Reply:
agree it's so much cleaner now (commit coming)